modules/in_band_dtmf/in_band_dtmf.c

```diff
@@ -0,0 +1,329 @@
+/**
+ * @file in_band_dtmf.c  DTMF decoder
+ */
+#include <time.h>
+#include <ctype.h>
+#include <string.h>
+#include <re.h>
+#include <rem.h>
+#include <baresip.h>
+
+
+/**
+ * @defgroup in_band_dtmf in_band_dtmf
+ *
+ * Audio filter that de- and encodes DTMF tones.
+ *
+ * The tone length of an encoded tone is 0.1s.
+ * New tones can be added while encoding is active.
+ */
+
+
+struct in_band_dtmf_filt_dec {
+	struct aufilt_enc_st af;  /* inheritance */
+	struct dtmf_dec *dec;
+	struct ua *ua;
+	struct call *call;
+	struct le le_priv;
+};
+
+
+struct in_band_dtmf_filt_enc {
+	struct aufilt_enc_st af;  /* inheritance */
+	struct mbuf *mb;
+	unsigned srate;
+	struct le le_priv;
+};
+
+
+static struct list encs;
+static struct list decs;
+
+
+static void in_band_dtmf_dec_handler(char digit, void *arg)
+{
+	(void)arg;
+	char key_str[2];
+	struct in_band_dtmf_filt_dec *st;
+
+	key_str[0] = digit;
+	key_str[1] = '\0';
+
+	if (!list_count(&decs)) {
+		warning("in_band_dtmf: no active call\n");
+		return;
+	}
+
+	st = decs.head->data;
+
+	ua_event(st->ua, UA_EVENT_CALL_DTMF_START, st->call, "%s", key_str);
+	ua_event(st->ua, UA_EVENT_CALL_DTMF_END, st->call, NULL);
+}
+
+
+static void enc_destructor(void *arg)
+{
+	struct in_band_dtmf_filt_enc *st =
+			(struct in_band_dtmf_filt_enc *) arg;
+
+	list_unlink(&st->af.le);
+	list_unlink(&st->le_priv);
+	mem_deref(st->mb);
+}
+
+
+static void dec_destructor(void *arg)
+{
+	struct in_band_dtmf_filt_dec *st =
+			(struct in_band_dtmf_filt_dec *) arg;
+
+	list_unlink(&st->af.le);
+	list_unlink(&st->le_priv);
+	mem_deref(st->dec);
+}
+
+
+static int encode_update(struct aufilt_enc_st **stp, void **ctx,
+			 const struct aufilt *af, struct aufilt_prm *prm,
+			 const struct audio *au)
+{
+	struct in_band_dtmf_filt_enc *st;
+	int err = 0;
+	(void)ctx;
+	(void)af;
+	(void)au;
+
+	if (!stp || !prm)
+		return EINVAL;
+
+	st = mem_zalloc(sizeof(*st), enc_destructor);
+	if (!st)
+		return ENOMEM;
+
+	st->mb = mbuf_alloc(0);
+	if (!st->mb) {
+		mem_deref(st);
+		err = ENOMEM;
+	}
+	else {
+		st->srate = prm->srate;
+		list_append(&encs, &st->le_priv, st);
+		*stp = (struct aufilt_enc_st *)st;
+	}
+
+	return err;
+}
+
+
+static int encode(struct aufilt_enc_st *aufilt_enc_st, struct auframe *af)
+{
+	struct in_band_dtmf_filt_enc *st = (void *)aufilt_enc_st;
+	uint16_t* data = af->sampv;
+	uint16_t i;
+
+	if (mbuf_get_left(st->mb)) {
+		af->fmt = AUFMT_S16LE; /* TODO: Take care about format? */
```

Comment:
You can add checks to only support AUFMT_S16LE in this module.
The format conversion should be done by other modules, e.g. `auconv` ...

Reply:
Fixed by 2d1cc6d6